### PR TITLE
Update e2e 155

### PR DESCRIPTION
**Pull request by karenzheung**
Changes in this PR:

ProcessMessages function in message queue takes in context and exits if context is cancelled
e2e integration test runs chpl querier to populate healthit_product db
e2e integration test runs capability receiver to get capability statements from the queue

To test:
make test_e2e 
Log on to rabbitmq and check that test-queue has messages in it
test-queue should then be empty 

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: 
https://oncprojectracking.healthit.gov/support/browse/LANTERN-155
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [ ] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [ ] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [ ] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.

### CODE REVIEW

File: endpointmanager/pkg/capabilityhandler/capabilityreceiver.go [hunk 1]
```diff
@@ -195,7 +195,8 @@ func ReceiveCapabilityStatements(ctx context.Context,
 	}
 
 	errs := make(chan error)
-	go messageQueue.ProcessMessages(messages, saveMsgInDB, &args, errs)
+	ctx, _ = context.WithCancel(ctx)
```
Comment (edeyoung) [inline endpointmanager/pkg/capabilityhandler/capabilityreceiver.go:198]: since you're not using `cancel`, i think you can just pass in the original `ctx`

File: e2e/integration_tests/data_flow_test.go [hunk 2]
```diff
@@ -242,6 +245,123 @@ func Test_EndpointLinksAreAvailable(t *testing.T) {
...
+	client := &http.Client{
+		Timeout: time.Second * 35,
+	}
+	err = chplquerier.GetCHPLProducts(ctx, store, client)
+	failOnError(err)
+
+	healthit_prod_row = store.DB.QueryRow("SELECT COUNT(*) FROM healthit_products;")
+	err = healthit_prod_row.Scan(&hitp_count)
+	failOnError(err)
+	if hitp_count != expected_hitp_count {
```
Comment (edeyoung) [inline e2e/integration_tests/data_flow_test.go:269]: i got a failure here. i'm guessing chpl added in products since you wrote this.
i'd modify to `hitp_count < expected_hitp_count` and say that it's at least the number that you previously retrieved.

File: e2e/integration_tests/data_flow_test.go [hunk 3]
```diff
@@ -242,6 +245,123 @@ func Test_EndpointLinksAreAvailable(t *testing.T) {
...
+	failOnError(err)
+	var test_vendor_list []string
+	defer rows.Close()
+	for rows.Next() {
+		var vendor string
+		err = rows.Scan(&vendor)
+		failOnError(err)
+		test_vendor_list = append(test_vendor_list, vendor)
+	}
+	th.Assert(t, len(expected_vendor_list) == len(test_vendor_list), "Number of distinct vendors is not what was expected")
```
Comment (edeyoung) [inline e2e/integration_tests/data_flow_test.go:361]: i occasionally get this error:

image

this relies on actual responses and the timing of adding things to the queue, etc., so it's possible (at appears happens) that an endpoint with a particular vendor won't be received in time. perhaps just confirm that the most common vendor exists (Epic) (maybe two most common?) is in the list. This will show that the vendor mapping is being executed.

File: e2e/integration_tests/data_flow_test.go [hunk 4]
```diff
@@ -242,6 +245,124 @@ func Test_EndpointLinksAreAvailable(t *testing.T) {
...
+	var chID lanternmq.ChannelID
+	mq, chID, err = aq.ConnectToServerAndQueue(qUser, qPassword, qHost, qPort, qName)
+	defer mq.Close()
+	th.Assert(t, err == nil, err)
+	th.Assert(t, mq != nil, "expected message queue to be created")
+	th.Assert(t, chID != nil, "expected channel ID to be created")
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go capabilityhandler.ReceiveCapabilityStatements(ctx, store, mq, chID, qName)
+	time.Sleep(30 * time.Second)
```
Comment (bmath10) [inline e2e/integration_tests/data_flow_test.go:331]: It seems like context.WithDeadline or context.WithTimeout might make more sense here, but I am by no means a go expert. @edeyoung any thoughts?

**Comment by bmath10** [comment]
I got this error while running:
```
Error Trace:	data_flow_test.go:362
"[Epic Systems Corporation Allscripts Medical Information Technology, Inc. (MEDITECH)
Cerner Corporation CareEvolution, Inc.]" does not contain "[Epic Systems Corporation
Cerner Corporation]"
Test_RetrieveCapabilityStatements
Messages:   	List of distinct vendors should include Epic and Cerner
```
Any thoughts?

**Comment by edeyoung** [comment]
> I got this error while running:
> 
> ```
> Error Trace:	data_flow_test.go:362
> "[Epic Systems Corporation Allscripts Medical Information Technology, Inc. (MEDITECH)
> Cerner Corporation CareEvolution, Inc.]" does not contain "[Epic Systems Corporation
> Cerner Corporation]"
> Test_RetrieveCapabilityStatements
> Messages:   	List of distinct vendors should include Epic and Cerner
> ```
> 
> Any thoughts?

@bmath10 try pulling again. i was getting that error too, but the very last commit addressed it for me. it got added right before my approval.

**Comment by edeyoung** [comment]
@karenzheung the travis tests are failing because of this:

image

this tends to be due to an issue connecting with the chpl server. sometimes it's a cert issue. @mayerm94 may be able to help you debug this tomorrow.

File: e2e/integration_tests/data_flow_test.go (hunk 4 above)
Comment (edeyoung) [reply e2e/integration_tests/data_flow_test.go:331]: agree that `context.WithTimeout` is ideal.

**Comment by bmath10** [comment]
> > I got this error while running:
> > ```
> > Error Trace:	data_flow_test.go:362
> > "[Epic Systems Corporation Allscripts Medical Information Technology, Inc. (MEDITECH)
> > Cerner Corporation CareEvolution, Inc.]" does not contain "[Epic Systems Corporation
> > Cerner Corporation]"
> > Test_RetrieveCapabilityStatements
> > Messages:   	List of distinct vendors should include Epic and Cerner
> > ```
> > 
> > 
> > Any thoughts?
> 
> @bmath10 try pulling again. i was getting that error too, but the very last commit addressed it for me. it got added right before my approval.

That fixed it, thanks!

File: e2e/integration_tests/data_flow_test.go [hunk 5]
```diff
@@ -326,10 +331,9 @@ func Test_RetrieveCapabilityStatements(t *testing.T) {
 	th.Assert(t, mq != nil, "expected message queue to be created")
 	th.Assert(t, chID != nil, "expected channel ID to be created")
 
-	ctx, cancel := context.WithCancel(context.Background())
+	ctx, _ := context.WithTimeout(context.Background(), 30 * time.Second)
 	go capabilityhandler.ReceiveCapabilityStatements(ctx, store, mq, chID, qName)
 	time.Sleep(30 * time.Second)
```
Comment (edeyoung) [inline e2e/integration_tests/data_flow_test.go:336]: i believe that instead of doing the `time.Sleep` here, you'd want to wait for the context to be completed:

```golang
		select {
		case <-ctx.Done():
			// continue after context complete
		}
```

(untested)

I believe this code waits. you'd add the `default` flag if you didn't want it to wait.

File: e2e/integration_tests/data_flow_test.go [hunk 6]
```diff
@@ -346,6 +350,14 @@ func Test_RetrieveCapabilityStatements(t *testing.T) {
 	if fhir_version_count < 300 {
 		t.Fatalf("There should be at least 300 capability statement with fhir version specified")
 	}
+}
+
+func Test_VendorList(t *testing.T) {
+	var err error
+
+	if viper.GetString("chplapikey") == "" {
+		t.Skip("Skipping Test_GetCHPLProducts because the CHPL API key is not set.")
```
Comment (edeyoung) [inline e2e/integration_tests/data_flow_test.go:359]: update comment

**Comment by bmath10** [comment]
Looks good to me! I think all you have to do is update the .travis.yml file with the new env variable and you should be good!
